pkg/sql/backfill.go

```diff
@@ -886,6 +893,275 @@ func (sc *SchemaChanger) nRanges(
 	return len(rangeIds), nil
 }
 
+// TODO(adityamaru): Consider moving this to sql/backfill. It has a lot of
+// schema changer dependencies which will need to be passed around.
+func (sc *SchemaChanger) distIndexBackfill(
+	ctx context.Context,
+	version descpb.DescriptorVersion,
+	targetSpans []roachpb.Span,
+	filter backfill.MutationFilter,
+	indexBackfillBatchSize int64,
+) error {
+	inMemoryStatusEnabled := sc.execCfg.Settings.Version.IsActive(
+		ctx, clusterversion.VersionAtomicChangeReplicasTrigger)
+	readAsOf := sc.clock.Now()
+
+	// Variables to track progress of the index backfill.
+	origNRanges := -1
+	origFractionCompleted := sc.job.FractionCompleted()
+	fractionLeft := 1 - origFractionCompleted
+
+	// Index backfilling ingests SSTs that don't play nicely with running txns
+	// since they just add their keys blindly. Running a Scan of the target
+	// spans at the time the SSTs' keys will be written will calcify history up
+	// to then since the scan will resolve intents and populate tscache to keep
+	// anything else from sneaking under us. Since these are new indexes, these
+	// spans should be essentially empty, so this should be a pretty quick and
+	// cheap scan.
+	const pageSize = 10000
+	noop := func(_ []kv.KeyValue) error { return nil }
+	if err := sc.fixedTimestampTxn(ctx, readAsOf, func(ctx context.Context, txn *kv.Txn) error {
+		for _, span := range targetSpans {
+			// TODO(dt): a Count() request would be nice here if the target isn't
+			// empty, since we don't need to drag all the results back just to
+			// then ignore them -- we just need the iteration on the far end.
+			if err := txn.Iterate(ctx, span.Key, span.EndKey, pageSize, noop); err != nil {
+				return err
+			}
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// Gather the initial resume spans for the table.
+	var todoSpans []roachpb.Span
+	var mutationIdx int
+	if err := sc.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		var err error
+		todoSpans, _, mutationIdx, err = rowexec.GetResumeSpans(
+			ctx, sc.jobRegistry, txn, sc.execCfg.Codec, sc.descID, sc.mutationID, filter)
+		return err
+	}); err != nil {
+		return err
+	}
+
+	log.VEventf(ctx, 2, "indexbackfill: initial resume spans %+v", todoSpans)
+
+	var p PhysicalPlan
+	var evalCtx extendedEvalContext
+	var planCtx *PlanningCtx
+	// TODO(adityamaru): Is it okay to run the DistSQL flow and progress update
```

Comment:
Calling this out as a decision I took with potentially limited understanding of its implication.